main/inc/lib/webservices/Rest.php

```diff
@@ -1590,4 +1594,332 @@ private function encodeParams(array $additionalParams = [])
+            throw new Exception(get_lang('SessionNotRegistered'));
+        }
+
+        $promotionId = $modelSession['promotion_id'];
+        if ($promotionId) {
+            $sessionList = array_keys(SessionManager::get_all_sessions_by_promotion($promotionId));
+            $sessionList[] = $newSessionId;
+            SessionManager::subscribe_sessions_to_promotion($modelSession['promotion_id'], $sessionList);
+        }
+
+        $modelExtraFields = [];
+        $fields = SessionManager::getFilteredExtraFields($modelSessionId);
+        foreach($fields as $field) {
+            $modelExtraFields[$field['variable']] = $field['value'];
+        }
+        $allExtraFields = array_merge($modelExtraFields, $extraFields);
+        foreach($allExtraFields as $k => $v) {
+            if (!SessionManager::update_session_extra_field_value($newSessionId, $k, $v)) {
+                throw new Exception(get_lang('CouldNotUpdateExtraFieldValue'));
+            }
+        }
+
+        $courseList = array_keys(SessionManager::get_course_list_by_session_id($modelSessionId));
+        if (!SessionManager::add_courses_to_session($newSessionId, $courseList)) {
+            throw new Exception(get_lang('CoursesNotAddedToSession'));
+        }
+
+        if (api_is_multiple_url_enabled()) {
+            if (api_get_current_access_url_id() != -1) {
+                UrlManager::add_session_to_url(
+                    $newSessionId,
+                    api_get_current_access_url_id()
+                );
+            } else {
+                UrlManager::add_session_to_url($newSessionId, 1);
+            }
+        } else {
+            UrlManager::add_session_to_url($newSessionId, 1);
+        }
+
+        return [$newSessionId];
+    }
+
+    /**
+     * @param $sessionId
+     * @param $loginName
+     * @return array
+     * @throws Exception
+     */
+    public function subscribeUserToSessionFromUsername($sessionId, $loginName)
+    {
+        if (!SessionManager::isValidId($sessionId)) {
+            throw new Exception(get_lang('SessionNotFound'));
+        }
+
+        $userId = UserManager::get_user_id_from_username($loginName);
+        if (False === $userId) {
+            throw new Exception(get_lang('UserNotFound'));
+        }
+
+        $subscribed = SessionManager::subscribeUsersToSession($sessionId, [$userId], SESSION_VISIBLE_READ_ONLY, false);
+        if (!$subscribed) {
+            throw new Exception(get_lang('UserNotSubscribed'));
+        }
+
+        return [true];
+    }
+
+    /**
+     * @param $fieldName
+     * @param $fieldValue
+     * @return array
+     * @throws Exception
+     */
+    public function getSessionFromExtraField($fieldName, $fieldValue)
+    {
+        // find sessions that that have value in field
+        $valueModel = new ExtraFieldValue('session');
+        $sessionIdList = $valueModel->get_item_id_from_field_variable_and_field_value($fieldName, $fieldValue, false, false, true);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Line exceeds maximum limit of 120 characters; contains 131 characters](https://app.codacy.com/app/chamilo/chamilo-lms/pullRequest?prid=4554450)